### PR TITLE
add toolset and refactor data page

### PR DESCRIPTION
**Pull request by zindlerb**

### CODE REVIEW

File: client/js/components/chart.jsx [hunk 1]
```diff
@@ -92,16 +93,39 @@ var Chart = React.createClass({
...
       var xs = {};
+      var regressionDataLabel = "regression";
       xs[schemaName] = schemaName + "_x"; 
+      xs[regressionDataLabel] = regressionDataLabel + "_x";
+
+      var scatterPlotArray = transformCSVtoScatterPlot(csvData, schemaName, xIndex, yIndex);
+
+      var min = Math.min.apply(null, tplot[0]);
+      var max = Math.max.apply(null, tplot[0]);
+      /* jshint ignore:start */
```
Comment (dduugg) [inline client/js/components/chart.jsx]: ignore only the specific rule(s), e.g. /* jshint undef: false */

File: client/js/components/chart.jsx [hunk 2]
```diff
@@ -92,16 +93,39 @@ var Chart = React.createClass({
...
+      var regressionEquation = ss.linear_regression().data(graphPoints).line();
+      
+      var minRegressionY = regressionEquation(min);
+      var maxRegressionY = regressionEquation(max);
+      
+      scatterPlotArray.push([regressionDataLabel, minRegressionY, maxRegressionY], 
+                            [regressionDataLabel + "_x", min, max]);
+      /* jshint ignore:end */
```
Comment (dduugg) [inline client/js/components/chart.jsx]: shouldn't this come right after line 119? or am i missing something?

File: client/js/components/data_page.jsx [hunk 3]
```diff
@@ -107,27 +109,96 @@ var Table = React.createClass({
...
+  statics: {
+    getToolset: function(chartType, headerlessCSV, headerIndex, headerIndex2) {
+      
+      var columnArray = headerlessCSV.map(function(row) {
+        return parseInt(row[headerIndex], 10);
+      });
+
+      var toolset = {
+        Mean: Math.floor(ss.mean(columnArray)),
+        Median: ss.median(columnArray),
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: Do you need `Math.floor` anywhere else (e.g., median of an even-length array)?

File: client/js/components/data_page.jsx [hunk 4]
```diff
@@ -3,6 +3,8 @@ var _ = require("lodash");
 var $ = require("jquery");
 var AppStore = require("../store/app_store");
 var Chart = require("./chart.jsx");
+var Papa = require("babyparse");
+var ss = require("simple-statistics");
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: var name isn't very helpful. i get that `simpleStatistics` is verbose, but even `stats` would be a huge improvement.

File: client/js/components/data_page.jsx [hunk 5]
```diff
@@ -107,27 +109,96 @@ var Table = React.createClass({
       chartType: "timeSeries"
     };
   },
+
+  statics: {
+    getToolset: function(chartType, headerlessCSV, headerIndex, headerIndex2) {
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: what is `headerIndex2`?

File: client/js/components/data_page.jsx [hunk 6]
```diff
@@ -107,27 +109,96 @@ var Table = React.createClass({
...
+    */
+    var getSchema = function(schemaName, callback) {
+      $.ajax({
+        url: "/api/schema/" + schemaName,
+        type: "GET",
+        beforeSend: function(request) {
+          request.setRequestHeader("x-jwt", localStorage.getItem("token"));
+        },
+        success: function(schemaData) {
+          //console.log("GET request for schema data successful.", schemaData);
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: please remove this

File: client/js/components/data_page.jsx [hunk 7]
```diff
@@ -3,6 +3,8 @@ var _ = require("lodash");
 var $ = require("jquery");
 var AppStore = require("../store/app_store");
 var Chart = require("./chart.jsx");
+var Papa = require("babyparse");
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: Papa?

File: client/js/components/data_page.jsx (hunk 3 above)
Comment (zindlerb) [inline client/js/components/data_page.jsx]: You are right, I should probably floor all.

File: client/js/components/data_page.jsx (hunk 4 above)
Comment (zindlerb) [inline client/js/components/data_page.jsx]: stats it is!

File: client/js/components/data_page.jsx [hunk 8]
```diff
@@ -107,27 +109,96 @@ var Table = React.createClass({
...
+  componentWillUnmount: function() {
+    $("#to-time-series").off("click");
+    $("#to-histogram").off("click");
+    $("#to-scatter-plot").off("click");
   },
   render: function() {
     var that = this;
     var setActiveHeader = function(newHeaderIndex) {
       that.setState({activeHeader: newHeaderIndex});
+      var headerlessCSV = _.rest(that.props.schemaCSVData);
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: var declarations should come first

File: client/js/components/data_page.jsx (hunk 5 above)
Comment (zindlerb) [inline client/js/components/data_page.jsx]: Woops from a diff feature branch. will remove.

File: client/js/store/app_store.js [hunk 9]
```diff
@@ -18,7 +18,7 @@ var AppStore = assign({}, EventEmitter.prototype, {
         request.setRequestHeader("x-jwt", localStorage.getItem("token"));
       },
       success: function(data) {
-        console.log("GET request for schema data successful.", data);
+        //console.log("GET request for schema data successful.", data);
```
Comment (dduugg) [inline client/js/store/app_store.js]: can we lose this line?

File: client/js/components/data_page.jsx (hunk 7 above)
Comment (zindlerb) [inline client/js/components/data_page.jsx]: haha will change to better name. :baby:

File: client/js/store/app_store.js [hunk 10]
```diff
@@ -102,6 +51,15 @@ var AppStore = assign({}, EventEmitter.prototype, {
     });
   },
 
+  updateCSVData: function(table) {
+    if (_.last(table).length === 1 && _.last(table)[0] === "") {
```
Comment (dduugg) [inline client/js/store/app_store.js]: extract `_.last(table)`, it's expensive

File: client/js/store/app_store.js [hunk 11]
```diff
@@ -102,6 +51,15 @@ var AppStore = assign({}, EventEmitter.prototype, {
     });
   },
 
+  updateCSVData: function(table) {
```
Comment (dduugg) [inline client/js/store/app_store.js]: `update` is a pretty broad term, this seems to do something very specific

File: client/js/components/data_page.jsx [hunk 12]
```diff
@@ -107,27 +109,95 @@ var Table = React.createClass({
...
+      var chartType = chartEvent[1];
+      
+      $(document).on("click", chartID, function() {
+        that.setState({chartType: chartType});  
+      });
+    });
+
+  },
+  componentWillUnmount: function() {
+    $("#to-time-series").off("click");
```
Comment (dduugg) [inline client/js/components/data_page.jsx]: can you give these a common class instead? e.g. `.to-chart`

File: client/js/components/chart.jsx [hunk 13]
```diff
@@ -125,16 +151,19 @@ var Chart = React.createClass({
...
         }
       });
+    },
+    renderVisualization: function(chartType, csvData, activeHeader, schemaName) {
+      if (chartType === "histogram") {
+        Chart.renderHistogramChart(csvData, activeHeader);
+      } else if (chartType === "timeSeries") {
+        Chart.renderTimeSeriesChart(csvData, activeHeader);
+      } else if (chartType === "scatterPlot") {
+        Chart.renderScatterplotChart(csvData, schemaName, activeHeader, 3);
```
Comment (dduugg) [inline client/js/components/chart.jsx]: what is `3`?

**Comment by cmperez** [comment]
:+1: after dug comments addressed

File: client/js/components/chart.jsx (hunk 13 above)
Comment (zindlerb) [inline client/js/components/chart.jsx]: Scatterplots require two data points. Currently we only have ui artifacts to select one data point so 3 is the default index of the second data point. I will later add the ability for the user to choose the index of the second data point.

Comment (dduugg) [inline client/js/components/chart.jsx]: ok cool, there should be a TODO here but no biggie

**Comment by dduugg** [comment]
the jshint is still unaddressed but good enough for me. merging.

**Pull request merged**